packages/fuels-core/src/code_gen/abigen.rs

```diff
@@ -110,48 +110,50 @@ impl Abigen {
                     }
 
                     impl #name {
-                        #contract_functions
+                        pub fn new(contract_id: String, wallet: WalletUnlocked) -> Self {
+                            let contract_id = Bech32ContractId::from_str(&contract_id).expect("Invalid contract id");
+                            Self { contract_id, wallet }
+                        }
 
-                        pub fn _get_contract_id(&self) -> &Bech32ContractId {
+                        pub fn contract_id(&mut self, contract_id: String) -> &mut Self {
+                            self.contract_id = Bech32ContractId::from_str(&contract_id).expect("Invalid contract id");
+                            self
+                        }
+
+                        pub fn wallet(&mut self, wallet: WalletUnlocked) -> &mut Self {
+                            self.wallet = wallet;
+                            self
+                        }
+
+                        pub fn get_contract_id(&self) -> &Bech32ContractId {
                             &self.contract_id
                         }
 
-                        pub fn _get_wallet(&self) -> WalletUnlocked {
+                        pub fn get_wallet(&self) -> WalletUnlocked {
                             self.wallet.clone()
                         }
 
-                        pub fn _with_wallet(&self, mut wallet: WalletUnlocked) -> Result<Self, SDKError> {
+                        pub fn with_wallet(&self, mut wallet: WalletUnlocked) -> Result<Self, SDKError> {
                            let provider = self.wallet.get_provider()?;
                            wallet.set_provider(provider.clone());
 
                            Ok(Self { contract_id: self.contract_id.clone(), wallet: wallet })
                         }
+
+                        pub fn methods(&self) -> #methods_name {
+                            #methods_name { contract_id: self.contract_id.clone(), wallet: self.wallet.clone() }
```

Comment:
Should this have a reference to the instance instead? What should the user expect to happen if they change the wallet of the instance with the  above `wallet()` method.

Reply:
They must call `methods()` again. AFAIK rust will not allow to have a reference to a dropped wallet. As the new one would replace it.

Reply:
This creates a pitfall. It's not evident that if you change the wallet, the methods are affected too. 
Can we use a reference to the actual instance or add a warning to the docs?

Reply:
(I am writing this without trying it out).  `wallet()` and `with_wallet` both create  a new instance so I think the reference is again no go. IMO documentation should be enough.

Reply:
I implemented the second option.